File: app/src/main/java/com/popalay/cardme/data/repositories/CardRepository.java
package com.popalay.cardme.data.repositories;

import android.content.Context;

import com.github.popalay.rxrealm.RxRealm;
import com.popalay.cardme.data.models.Card;
import com.popalay.cardme.data.models.Holder;

import java.util.List;

import javax.inject.Inject;
import javax.inject.Singleton;

import io.realm.Sort;
import rx.Completable;
import rx.Observable;
import rx.Single;

@Singleton
public class CardRepository {

    private final Context context;

    @Inject
    public CardRepository(Context context) {
        this.context = context;
    }

    public Completable save(Card card) {
        return Completable.fromAction(() -> RxRealm.generateObjectId(card, (realm, id) -> {
            if (card.getId() > 0) {
                card.setId(id);
            }
            final Holder realmHolder = realm.where(Holder.class).equalTo(Holder.NAME, card.getHolder().getName())
                    .findFirst();
            if (realmHolder != null) {
                card.setHolder(realmHolder);
            } else {
                final Number num = realm.where(Holder.class).max(Holder.ID);
                final long nextID = num != null ? num.longValue() + 1L : 0L;
                card.getHolder().setId(nextID);
            }
            card.getHolder().setCardsCount(card.getHolder().getCardsCount() + 1);
            realm.copyToRealmOrUpdate(card);
        }));
    }

    public Observable<List<Card>> getAll() {
        return RxRealm.listenList(realm -> realm.where(Card.class)
                .findAllSorted(Card.ID, Sort.DESCENDING, Card.USAGE, Sort.DESCENDING)
                .sort(Card.USAGE, Sort.DESCENDING));
    }

    public Observable<List<Card>> getAllByHolder(long holderId) {
        return RxRealm.listenList(realm -> realm.where(Card.class)
                .equalTo(Card.HOLDER_ID, holderId)
                .findAllSorted(Card.ID, Sort.DESCENDING)
                .sort(Card.USAGE, Sort.DESCENDING));
    }

    public Completable incCardUsage(final Card card) {
        return Completable.fromAction(() -> RxRealm.doTransactional(realm -> {
            card.setUsage(card.getUsage() + 1);
            realm.copyToRealmOrUpdate(card);
        }));
    }

    public Completable remove(final Card card) {
        return Completable.fromAction(() -> RxRealm.doTransactional(realm -> {
            realm.where(Card.class).equalTo(Card.ID, card.getId()).findAll().deleteAllFromRealm();
        }));
    }

    public Single<Integer> countByHolder(Holder holder) {
        return RxRealm.getList(realm -> realm.where(Card.class).equalTo(Card.HOLDER_ID, holder.getId()).findAll())
                .map(List::size);
    }
}

File: data/src/main/kotlin/com/popalay/cardme/data/repositories/CardRepository.kt
package com.popalay.cardme.data.repositories

import com.google.gson.Gson
import com.popalay.cardme.data.dao.CardDao
import com.popalay.cardme.data.models.Card
import io.reactivex.Completable
import io.reactivex.Flowable
import io.reactivex.Single
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class CardRepository @Inject constructor(
        private val cardDao: CardDao,
        private val gson: Gson
) {

    fun get(cardNumber: String): Flowable<Card> = Flowable.empty()/*RxRealm.listenElement {
        it.where(Card::class.java).equalTo(Card.NUMBER, cardNumber).findAll()
    }*/

    fun update(cards: List<Card>): Completable = Completable.complete()
    /*RxRealm.doTransactional { it.copyToRealmOrUpdate(cards) }*/

    fun getAll(): Flowable<List<Card>> = cardDao.getAllNotTrashed()

    fun getAllTrashed(): Flowable<List<Card>> = Flowable.empty()/*RxRealm.listenList {
        it.where(Card::class.java)
                .equalTo(Card.IS_TRASH, true)
                .findAllSorted(Card.POSITION, Sort.ASCENDING, Card.HOLDER_NAME, Sort.ASCENDING)
    }*/

    fun markAsTrash(card: Card): Completable = Completable.complete()/*RxRealm.doTransactional {
        it.where(Card::class.java).equalTo(Card.NUMBER, card.number).findFirst().isTrash = true
    }*/

    fun removeTrashed(): Completable = Completable.complete()/*RxRealm.doTransactional {
        it.where(Card::class.java).equalTo(Card.IS_TRASH, true).findAll().deleteAllFromRealm()
    }*/

    fun restore(card: Card): Completable = Completable.complete()/*RxRealm.doTransactional {
        it.where(Card::class.java).equalTo(Card.NUMBER, card.number).findFirst().isTrash = false
    }
*/
    fun cardIsNew(cardNumber: String): Single<Boolean> = Single.just(true)/*RxRealm.getElement {
        it.where(Card::class.java)
                .equalTo(Card.IS_TRASH, false)
                .equalTo(Card.NUMBER, cardNumber).findFirst()
    }.isEmpty*/

    fun prepareForSharing(card: Card): Single<String> = Single.fromCallable { gson.toJson(card) }

    fun getFromJson(source: String): Single<Card> = Single.fromCallable {
        gson.fromJson<Card>(source, Card::class.java)
    }

}
